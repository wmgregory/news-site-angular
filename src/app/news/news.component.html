<h2>News</h2>

<div class="container-fluid">
  <div class="row">

    <div class="card-columns">

      <div class="card" *ngFor="let article of (articles$ | async)">
        <img class="card-img-top" *ngIf="article.urlToImage" [src]="article.urlToImage" [alt]="article.title" />

        <div class="card-body">
          <h5 class="card-title">{{ article.title }}</h5>
          <p class="card-text">{{ article.description }}</p>
          <a class="btn btn-primary" [href]="article.url">Go somewhere</a>
        </div>
      </div>

    </div>

  </div>
</div>
