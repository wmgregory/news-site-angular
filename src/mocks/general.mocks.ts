export const httpMock = {
  get: jest.fn(),
  post: jest.fn(),
};
