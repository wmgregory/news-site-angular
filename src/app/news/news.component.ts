import { Component, OnInit } from '@angular/core';

import { Observable } from 'rxjs/Observable';

import { Article } from '../models/Article.model';
import { NewsService } from '../services/news.service';

@Component({
  selector: 'app-news',
  templateUrl: './news.component.html',
  styleUrls: ['./news.component.scss']
})
export class NewsComponent implements OnInit {

  public articles$: Observable<Article[]>;

  constructor(private newsService: NewsService) {}

  ngOnInit() {
    this.articles$ = this.getArticles();
  }

  getArticles(): Observable<Article[]> {
    return this.newsService.getArticles();
  }

  searchKeywords(keywords?: string): void {
    this.articles$ = this.newsService.getArticles(keywords);
  }
}
