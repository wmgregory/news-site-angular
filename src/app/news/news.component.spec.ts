import { CUSTOM_ELEMENTS_SCHEMA } from '@angular/core';
import { async, ComponentFixture, TestBed } from '@angular/core/testing';

import { Observable } from 'rxjs/Observable';
import 'rxjs/add/observable/of';

import { NewsComponent } from './news.component';

import { NewsService } from '../services/news.service';
import { newsServiceMock } from '../services/news.service.mocks';

describe('NewsComponent', () => {
  let component: NewsComponent;
  let fixture: ComponentFixture<NewsComponent>;

  const articlesMock = [
      {
        'source': {'id': 'techcrunch', 'name': 'TechCrunch'},
        'author': 'Jon Evans',
        'title': 'Giving up on \'diversity and inclusion\'',
        'description': 'I went to see Arlan Hamilton of Backstage Capital speak this week. Her remarkable story is pretty well known by nowâ€¦',
        'url': 'https://techcrunch.com/2018/04/29/giving-up-on-diversity-and-inclusion/',
        'urlToImage': 'https://techcrunch.com/wp-content/uploads/2018/04/shoe-diversity.jpg?w=764',
        'publishedAt': '2018-04-29T13:00:12Z'
      }
  ];

  beforeEach(async(() => {
    TestBed.configureTestingModule({
      declarations: [ NewsComponent ],
      providers: [
        { provide: NewsService, useValue: newsServiceMock },
      ],
      schemas: [ CUSTOM_ELEMENTS_SCHEMA ],
    })
    .compileComponents();

    jest.spyOn(newsServiceMock, 'getArticles').mockReturnValue(Observable.of(articlesMock));

  }));

  beforeEach(() => {

    fixture = TestBed.createComponent(NewsComponent);
    component = fixture.componentInstance;

    fixture.detectChanges();
  });

  it('should create', () => {
    expect(component).toBeTruthy();
    expect(fixture).toMatchSnapshot();
  });

  it('should contain an article', async () => {
    const articles = await component.articles$;
    expect(articles.value).toEqual(articlesMock);
  });

  it('should search articles', async () => {
    const keywords = ['news'];
    const newArticlesMock = [...articlesMock, ...articlesMock];
    jest.spyOn(newsServiceMock, 'getArticles').mockReturnValue(Observable.of(newArticlesMock));

    component.searchArticles(keywords);
    const articles = await component.articles$;

    expect(newsServiceMock.getArticles).toHaveBeenCalledWith(keywords);
    expect(articles.value).toEqual(newArticlesMock);
  });
});
