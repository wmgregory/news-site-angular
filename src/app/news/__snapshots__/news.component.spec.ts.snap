// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`NewsComponent should create 1`] = `
<app-news
  articles$={[Function ScalarObservable]}
  newsService={[Function Object]}
>
  <h2>
    News
  </h2>
  
  <div
    class="container-fluid"
  >
    
    
    <div
      class="row"
    >
      
      
      <app-search />
      
    
    </div>
    
  
    
    <div
      class="row"
    >
      
  
      
      <div
        class="card-columns"
      >
        
  
        
        
        <div
          class="card"
        >
          
          
          
          <img
            alt="Giving up on 'diversity and inclusion'"
            class="card-img-top"
            src="https://techcrunch.com/wp-content/uploads/2018/04/shoe-diversity.jpg?w=764"
          />
          
  
          
          <div
            class="card-body"
          >
            
            
            <h5
              class="card-title"
            >
              Giving up on 'diversity and inclusion'
            </h5>
            
            
            <p
              class="card-text"
            >
              I went to see Arlan Hamilton of Backstage Capital speak this week. Her remarkable story is pretty well known by nowâ€¦
            </p>
            
            
            <a
              class="btn btn-primary"
              href="https://techcrunch.com/2018/04/29/giving-up-on-diversity-and-inclusion/"
            >
              Go somewhere
            </a>
            
          
          </div>
          
        
        </div>
        
  
      
      </div>
      
  
    
    </div>
    
  
  </div>
  
</app-news>
`;
